mlx/backend/metal/reduce.cpp

```diff
@@ -36,37 +41,88 @@ void all_reduce_dispatch(
     MTL::ComputeCommandEncoder* compute_encoder,
     metal::Device& d) {
   // Get kernel and encode buffers
-  size_t in_size = in.size();
-  auto kernel = d.get_kernel("all_reduce_" + op_name + type_to_name(in));
 
-  compute_encoder->setComputePipelineState(kernel);
-  set_array_buffer(compute_encoder, in, 0);
-  set_array_buffer(compute_encoder, out, 1);
-  compute_encoder->setBytes(&in_size, sizeof(size_t), 2);
+  Dtype out_dtype = out.dtype();
+  bool is_out_64b_int = is_64b_int(out_dtype);
+  std::string kernel_name =
+      (is_out_64b_int) ? "all_reduce_no_atomics_" : "all_reduce_";
+  kernel_name += (op_name + type_to_name(in));
 
-  // Set grid dimensions
+  auto kernel = d.get_kernel(kernel_name);
 
   // We make sure each thread has enough to do by making it read in
   // at least n_reads inputs
   int n_reads = REDUCE_N_READS;
 
+  size_t in_size = in.size();
+  size_t intermediate_size;
+
   // mod_in_size gives us the groups of n_reads needed to go over the entire
   // input
   uint mod_in_size = (in_size + n_reads - 1) / n_reads;
-
   NS::UInteger thread_group_size = kernel->maxTotalThreadsPerThreadgroup();
   thread_group_size =
       mod_in_size > thread_group_size ? thread_group_size : mod_in_size;
+  uint simd_size = kernel->threadExecutionWidth();
+  thread_group_size =
+      ((thread_group_size + simd_size - 1) / simd_size) * simd_size;
 
   // If the number of thread groups needed exceeds 1024, we reuse threads groups
   uint n_thread_groups = safe_div(mod_in_size, thread_group_size);
   n_thread_groups = std::min(n_thread_groups, 1024u);
   uint nthreads = n_thread_groups * thread_group_size;
 
+  // Allocate intermediate array to store partial reduction results from all
+  // thread groups
+  array* output = &out;
+
+  // Allocate intermediate array to store partial reduction results
+  std::unique_ptr<array> intermediate;
+  if (is_out_64b_int) {
+    std::vector<array> inputs = {};
+    intermediate_size = n_thread_groups;
+    std::vector<int> intermediate_shape = {static_cast<int>(intermediate_size)};
+    intermediate =
+        std::make_unique<array>(intermediate_shape, out_dtype, nullptr, inputs);
+    intermediate->set_data(allocator::malloc_or_wait(intermediate->nbytes()));
+    output = intermediate.get();
+  }
+
+  // First compute dispatch
+  compute_encoder->setComputePipelineState(kernel);
+  set_array_buffer(compute_encoder, in, 0);
+  set_array_buffer(compute_encoder, *output, 1);
+  compute_encoder->setBytes(&in_size, sizeof(size_t), 2);
+
   MTL::Size group_dims = MTL::Size(thread_group_size, 1, 1);
   MTL::Size grid_dims = MTL::Size(nthreads, 1, 1);
 
   compute_encoder->dispatchThreads(grid_dims, group_dims);
+
+  if (is_out_64b_int) {
```

Comment:
I think this would simplify quite a lot if it was clearly split up in two separate cases. Normal and atomic free. The normal one should be the same as before and for the atomic free you would allocate a new array and call the kernel twice one in with the intermediate and one with the out.

Importantly, the second call is always a single threadgroup! Otherwise there is a bug. So no need for the thread group math. You could check that it indeed fits in a single threadgroup if you want for extra safety. Also the 2nd call can be avoided if the input array is smaller than `1024 * N_READS` which is quite a few cases.

One more important thing is that the intermediate array is currently freed as soon as the function exits (which is before the kernel is run). The way we deal with intermediate arrays is we hold a copy in a completion handler and we clear it so that it gets freed after the kernel is run. See https://github.com/ml-explore/mlx/blob/65d0b8df9fb02a30445fc5431aa2d8e73ca0af2f/mlx/backend/metal/matmul.cpp#L251 or https://github.com/ml-explore/mlx/blob/65d0b8df9fb02a30445fc5431aa2d8e73ca0af2f/mlx/backend/metal/softmax.cpp#L80 for examples.

Reply:
> I think this would simplify quite a lot if it was clearly split up in two separate cases. Normal and atomic free. The normal one should be the same as before and for the atomic free you would allocate a new array and call the kernel twice one in with the intermediate and one with the out.

Sure, I can do that. But I think there's going to some amount of duplication if we go down that path.

> Importantly, the second call is always a single threadgroup! Otherwise there is a bug. So no need for the thread group math. You could check that it indeed fits in a single threadgroup if you want for extra safety. Also the 2nd call can be avoided if the input array is smaller than 1024 * N_READS which is quite a few cases.

Ack, that's a good catch. I'll address that.

> One more important thing is that the intermediate array is currently freed as soon as the function exits (which is before the kernel is run).
Ack. I'm guessing this is for required so the reuse allocator in MLX doesn't recycle it until prior use is complete (since the underlying Metal resource should be retained by the driver once we bind it.) I'll add a completion handler block .